dbms/src/Columns/ColumnString.cpp

```diff
@@ -527,42 +581,158 @@ void ColumnString::countSerializeByteSizeForColumnArray(
+                - offsetAt(array_offsets[i - 1]);
+    }
+}
+
+void ColumnString::serializeForCmpToPos(
+    PaddedPODArray<char *> & pos,
+    size_t start,
+    size_t length,
+    bool has_null,
+    const TiDB::TiDBCollatorPtr & collator,
+    String * sort_key_container) const
+{
+    if (has_null)
+    {
+        if likely (collator != nullptr)
+            serializeToPosImpl</*has_null=*/true, /*has_collator=*/true>(
+                pos,
+                start,
+                length,
+                collator,
+                sort_key_container);
+        else
+            serializeToPosImpl</*has_null=*/true, /*has_collator=*/false>(pos, start, length, nullptr, nullptr);
+    }
+    else
+    {
+        if likely (collator != nullptr)
+            serializeToPosImpl</*has_null=*/false, /*has_collator=*/true>(
+                pos,
+                start,
+                length,
+                collator,
+                sort_key_container);
+        else
+            serializeToPosImpl</*has_null=*/false, /*has_collator=*/false>(pos, start, length, nullptr, nullptr);
+    }
 }
 
 void ColumnString::serializeToPos(PaddedPODArray<char *> & pos, size_t start, size_t length, bool has_null) const
 {
     if (has_null)
-        serializeToPosImpl<true>(pos, start, length);
+        serializeToPosImpl</*has_null=*/true, /*has_collator=*/false>(pos, start, length, nullptr, nullptr);
     else
-        serializeToPosImpl<false>(pos, start, length);
+        serializeToPosImpl</*has_null=*/false, /*has_collator=*/false>(pos, start, length, nullptr, nullptr);
 }
 
-template <bool has_null>
-void ColumnString::serializeToPosImpl(PaddedPODArray<char *> & pos, size_t start, size_t length) const
+template <bool has_null, bool has_collator>
+void ColumnString::serializeToPosImpl(
+    PaddedPODArray<char *> & pos,
+    size_t start,
+    size_t length,
+    const TiDB::TiDBCollatorPtr & collator,
+    String * sort_key_container) const
 {
     RUNTIME_CHECK_MSG(length <= pos.size(), "length({}) > size of pos({})", length, pos.size());
     RUNTIME_CHECK_MSG(start + length <= size(), "start({}) + length({}) > size of column({})", start, length, size());
+    if constexpr (has_collator)
+        RUNTIME_CHECK(collator && sort_key_container);
 
-    /// countSerializeByteSize has already checked that the size of one element is not greater than UINT32_MAX
+    /// countSerializeForCmpByteSize has already checked that the size of one element is not greater than UINT32_MAX
     for (size_t i = 0; i < length; ++i)
     {
         if constexpr (has_null)
         {
             if (pos[i] == nullptr)
                 continue;
         }
+
         UInt32 str_size = sizeAt(start + i);
+        const void * src = &chars[offsetAt(start + i)];
+        if constexpr (has_collator)
+        {
+            auto sort_key = collator->sortKey(reinterpret_cast<const char *>(src), str_size - 1, *sort_key_container);
```

Comment:
The `sortKey` is a virtual function. How about adding a batch version to reduce the overhead of virtual functions?

Reply:
Add some marcos so we can static_cast collator to Derived class and call method in specific Derived class.